neps/nep-0393.md

```diff
@@ -0,0 +1,480 @@
+These functions must relay calls to an SBT registry, which will emit appropriate events.
+We recommend that the all functions related to an event will take an optional `memo: Option<String>` argument for accounting purposes.
+
+```rust
+trait SBT {
+    /// the function should overwrite `metadata.class = class`.
+    /// Must provide enough NEAR to cover registry storage cost.
+    // #[payable]
+    fn sbt_mint(
+        &mut self,
+        account: AccountId,
+        class: u64,
+        metadata: TokenMetadata,
+        memo: Option<String>,
+    ) -> TokenId;
+
+    /// Creates a new, unique token and assigns it to the `receiver`.
+    /// `token_spec` is a vector of pairs: owner AccountId and TokenMetadata.
+    /// Must provide enough NEAR to cover registry storage cost.
+    // #[payable]
+    fn sbt_mint_multi(
+        &mut self,
+        token_spec: Vec<(AccountId, TokenMetadata)>,
+        memo: Option<String>,
+    ) -> Vec<TokenId>;
+
+    // #[payable]
+    fn sbt_recover(&mut self, from: AccountId, to: AccountId, memo: Option<String>);
+
+    fn sbt_renew(&mut self, tokens: Vec<TokenId>, expires_at: u64, memo: Option<String>);
+
+    fn sbt_revoke(token: Vec<u64>, memo: Option<String>) -> bool;
+}
+```
+
+## Reference Implementation
+
+- Common [type definitions](https://github.com/alpha-fi/i-am-human/tree/master/contracts/sbt) (events, traits).
+- https://github.com/alpha-fi/i-am-human/tree/master/contracts/soulbound
+
+## Example Flows
+
+```mermaid
+sequenceDiagram
+    actor Alice
+    actor Issuer1 as SBT_1 Issuer
+
+    participant SBT1 as SBT_1 Contract
+    participant SBT_Registry
+
+    Issuer1->>SBT1: sbt_mint(alice, 1, metadata)
+    activate SBT1
+    SBT1-)SBT_Registry: sbt_mint([[alice, metadata])
+    SBT_Registry->>SBT_Registry: emit Mint(SBT_1_Contract, alice, [238])
+    SBT_Registry-)SBT1: [238]
+    deactivate SBT1
+
+    Note over Alice,SBT_Registry: now Alice can query registry to check her SBT
+
+    Alice-->>SBT_Registry: sbt(SBT_1_Contract, 238)
+    SBT_Registry-->>Alice: {token: 238, owner: alice, metadata}
+
+    Note over Alice,SBT_Registry: Alice participates in other community and applies for another SBT.<br/> She will get 2 SBTs. However she uses a different wallet: alice2.
+
+    participant SBT2 as SBT_2 Contract
+    actor Issuer2 as SBT_2 Issuer
+
+    Issuer2->>SBT2: sbt_mint_multi([[alice2, metadata2], [alice2, metadata3]])
+    activate SBT2
+    SBT2-)SBT_Registry: sbt_mint([[alice2, metadata2], [alice2, metadata3]])
+    SBT_Registry->>SBT_Registry: emit Mint(SBT_2_Contract, alice2, [7991, 17992])
+    SBT_Registry-)SBT2: [7991, 1992]
+    deactivate SBT2
+
+    Alice-->>SBT_Registry: sbt(SBT_2_Contract, 7991)
+    SBT_Registry-->>Alice: {token: 7991, owner: alice2, metadata: metadata2}
+
+    Note over Alice,SBT_Registry: After a while Alice decides to merge her `alice2` account into `alice2`
+    Alice->>SBT_Registry: sbt_soul_transfer(alice) --accountId alice2
```

Comment:
The notation here is not great. No other flows have specified which wallet Alice is using. I suggest having either two agents, as Alice1 and Alice2, or using a notation showing which of the accounts triggered which actions (both for `sbt_mint` and `sbt_mint_multi`).

Reply:
This is an example of one agent (Alice) who has 2 accounts and is clearly displayed by the diagram. 
(you should look at the diagram). If you are confused with `alice` being an account name, then I can rename it to `wallet1`.

Reply:
No, I have looked at the diagram. And no, it is not clear. It may be clear to you, as you made it, and know the exact intent, but it is not serving its purpose of helping to explain the ideas better. Perhaps if the two SBT issuances were also triggered by Alice, then you can introduce this notation early, so it's more clear later.

Reply:
I have updated the diagram and split it into 3 diagrams. Hope it is more clear now. Thanks for asking about it.

Reply:
Yes, it is much better. Thank you!